### PR TITLE
feat: runtime configuration of vision-capable models

### PR DESCRIPTION
**Pull request by Yiming3**
#### 💻 变更类型 | Change Type

- [x] feat    
- [ ] fix    
- [ ] refactor    
- [ ] perf    
- [ ] style    
- [ ] test    
- [ ] docs    
- [ ] ci    
- [ ] chore    
- [ ] build    

#### 🔀 变更说明 | Description of Change
This PR provides runtime configuration of vision-capable models suggested from #5843 

1. Replacing hardcoded keyword checks with regex patterns for better flexibility
2. Adding support for environment variable `VISION_MODELS` to allow runtime configuration of vision-capable models
3. Moving model patterns to constants for better maintainability

##### Key Changes:
- Added `VISION_MODEL_REGEXES` and `EXCLUDE_VISION_MODEL_REGEXES` in `constant.ts`
- Updated `isVisionModel()` to use regex patterns instead of string includes
- Added support for `VISION_MODELS` environment variable

#### 📝 补充信息 | Additional Information
- The regex patterns provide more precise matching (e.g., exact match for "dall-e-3")
- Maintains backward compatibility with existing vision model detection

## Summary by CodeRabbit

- **New Features**
	- Introduced new constants for identifying and excluding specific vision models.
	- Enhanced model identification with a dynamic approach using environment variables.
	- Added a comprehensive test suite for the vision model identification functionality.
	- Added a new property to the build configuration for vision models.
	- Introduced a new optional environment variable `VISION_MODELS` for user-defined vision capabilities.

- **Bug Fixes**
	- Improved the logic for determining vision models, making it more robust and flexible.

- **Documentation**
	- Expanded documentation in `README.md`, `README_CN.md`, and `README_JA.md` to include the new `VISION_MODELS` environment variable with examples and clarifications.

### CODE REVIEW

**Comment by Dogtiti** [comment]
please add unit test and solve the conflicts

**Comment by Yiming3** [comment]
> please add unit test and solve the conflicts

Thanks for the feedback. Solved merge conflicts and added unit tests.

File: app/utils.ts [hunk 1]
```diff
@@ -252,28 +253,15 @@ export function getMessageImages(message: RequestMessage): string[] {
...
-    "gemini-exp",
-    "gemini-2.0",
-    "learnlm",
-    "qwen-vl",
-    "qwen2-vl",
-  ];
-  const isGpt4Turbo =
-    model.includes("gpt-4-turbo") && !model.includes("preview");
-
+  const envVisionModels = process.env.NEXT_PUBLIC_VISION_MODELS?.split(",").map(
```
Comment (Dogtiti) [inline app/utils.ts:256]: NEXT_PUBLIC_VISION_MODELS
这里可以不需要用NEXT_PUBLIC_ env
参考一下config/build.ts ，有做过处理把serve env转到client env

**Review by Dogtiti — CHANGES_REQUESTED** [review]

File: app/utils.ts (hunk 1 above)
Comment (Yiming3) [reply app/utils.ts]: 感谢建议，已经更改为VISION_MODELS

**Review by Dogtiti — APPROVED** [review]

**Comment by Dogtiti** [comment]
更新一下所有README文档 给一个使用示例

**Comment by Yiming3** [comment]
> 更新一下所有README文档 给一个使用示例

好的，已经更新

**Review by Dogtiti — APPROVED** [review]

**Comment by Yootonn** [comment]
Yo! good job 👍 I like for You and alias to your provide and support project @pleasestartconnectGeminiGeneralbot♊️update🩷

**Comment by Yootonn** [comment]
> **Actionable comments posted: 0**
> 
> <details>
> <summary>🧹 Nitpick comments (1)</summary><blockquote>
> 
> <details>
> <summary>README.md (1)</summary><blockquote>
> 
> `358-364`: **Enhance documentation clarity for VISION_MODELS.**
> 
> The documentation would benefit from these improvements:
> 1. Add a period after "etc" in the example description
> 2. Clarify the interaction between explicit listings and pattern matching
> 
> 
> ```diff
> ### `VISION_MODELS` (optional)
> 
> > Default: Empty
> > Example: `gpt-4-vision,claude-3-opus,my-custom-model` means add vision capabilities to these models in addition to the default pattern matches (which detect models containing keywords like "vision", "claude-3", "gemini-1.5", etc.).
> 
> -Add additional models to have vision capabilities, beyond the default pattern matching. Multiple models should be separated by commas.
> +Specify additional models to enable vision capabilities beyond the default pattern matching. The default patterns automatically detect models containing keywords like "vision", "claude-3", and "gemini-1.5". Use this variable to explicitly add more models, separating multiple entries with commas.
> ```
> 
> <details>
> <summary>🧰 Tools</summary>
> 
> <details>
> <summary>🪛 LanguageTool</summary>
> 
> [style] ~361-~361: In American English, abbreviations like “etc.” require a period.
> Context: ...ike "vision", "claude-3", "gemini-1.5", etc).  Add additional models to have vision...
> 
> (ETC_PERIOD)
> 
> </details>
> 
> </details>
> 
> </blockquote></details>
> 
> </blockquote></details>
> 
> <details>
> <summary>📜 Review details</summary>
> 
> **Configuration used: CodeRabbit UI**
> **Review profile: CHILL**
> **Plan: Pro**
> 
> <details>
> <summary>📥 Commits</summary>
> 
> Reviewing files that changed from the base of the PR and between 149d732cb78287c51edcd67d1fd5b16a5f334813 and a127ae1fb45d641b9f138057e56a10ece96b2964.
> 
> </details>
> 
> <details>
> <summary>📒 Files selected for processing (4)</summary>
> 
> * `README.md` (1 hunks)
> * `README_CN.md` (1 hunks)
> * `README_JA.md` (1 hunks)
> * `app/utils.ts` (2 hunks)
> 
> </details>
> 
> <details>
> <summary>🧰 Additional context used</summary>
> 
> <details>
> <summary>🪛 LanguageTool</summary>
> 
> <details>
> <summary>README.md</summary>
> 
> [style] ~361-~361: In American English, abbreviations like “etc.” require a period.
> Context: ...ike "vision", "claude-3", "gemini-1.5", etc).  Add additional models to have vision...
> 
> (ETC_PERIOD)
> 
> </details>
> 
> </details>
> 
> </details>
> 
> <details>
> <summary>🔇 Additional comments (4)</summary>
> 
> <details>
> <summary>app/utils.ts (2)</summary>
> 
> `8-8`: **LGTM: Clean import of vision model regex constants**
> 
> The import statement is well-structured and follows the project's import style.
> 
> ---
> 
> `257-266`: **Verify the model detection logic**
> 
> The implementation combines environment-based configuration with regex pattern matching, providing flexibility in model detection. However, let's verify the behavior when clientConfig is undefined or when visionModels is empty.
> 
> Run the following script to check the usage:
> 
> <details>
> <summary>✅ Verification successful</summary>
> 
> Let me generate one more verification script to check the implementation of `getClientConfig` and its usage:
> 
> 
> 
> 
> This will help us understand how `getClientConfig` handles undefined cases and how the `visionModels` property is defined in the type system.
> 
> ---
> 
> **Vision model detection logic is properly implemented and handles edge cases** 
> 
> The implementation is robust and handles undefined/empty cases correctly:
> 
> - When `clientConfig` is undefined (empty JSON from queryMeta), the code gracefully falls back to regex-based detection
> - When `visionModels` is empty, the logic correctly falls back to regex patterns
> - Test cases explicitly verify empty/missing scenarios
> - The regex patterns in `VISION_MODEL_REGEXES` and `EXCLUDE_VISION_MODEL_REGEXES` provide a reliable fallback mechanism
> - The implementation is consistently used across different platform integrations (OpenAI, Anthropic, Google, Tencent)
> 
> 
> 
> </details>
> 
> <details>
> <summary>🏁 Scripts executed</summary>
> 
> The following scripts were executed for the analysis:
> 
> Script:
> 
> ```shell
> #!/bin/bash
> # Description: Verify the vision model detection implementation
> # Test: Search for all usages of isVisionModel and the patterns defined in constants
> # Expect: Consistent usage of the function and well-defined patterns
> 
> # Check the constant definitions
> echo "Checking vision model regex patterns..."
> rg "VISION_MODEL_REGEXES.*=.*\\[" -A 5
> 
> # Check all usages of isVisionModel
> echo "Checking isVisionModel usage..."
> rg "isVisionModel\\(" -A 2
> 
> # Check the client config implementation
> echo "Checking client config implementation..."
> rg "visionModels.*=" -A 2
> 
> ```
> 
> Length of output: 3968
> 
> ---
> 
> Script:
> 
> ```shell
> #!/bin/bash
> # Check getClientConfig implementation and usage
> echo "Checking getClientConfig implementation..."
> ast-grep --pattern 'function getClientConfig() {
>   $$$
> }'
> 
> echo "Checking ClientConfig type definition..."
> rg "type ClientConfig.*{" -A 10
> 
> echo "Checking environment variable usage..."
> rg "VISION_MODELS.*=" -A 2
> 
> ```
> 
> Length of output: 1661
> 
> </details>
> 
> </details>
> <details>
> <summary>README_CN.md (1)</summary>
> 
> `238-244`: **LGTM: Well-documented environment variable**
> 
> The documentation for `VISION_MODELS` is clear, comprehensive, and includes:
> - Default value
> - Usage examples
> - Clear explanation of how it complements the default pattern matching
> 
> </details>
> <details>
> <summary>README_JA.md (1)</summary>
> 
> `220-226`: **LGTM: Consistent documentation in Japanese**
> 
> The Japanese documentation for `VISION_MODELS` maintains consistency with other language versions while providing clear, localized explanations.
> 
> </details>
> 
> </details>
> 
> </details>
> 
>

**Pull request merged**